File: backend/src/repositories/EventRepository.ts
import { DatabaseManager } from '../db/DatabaseManager';
import { Event, EventStatus } from '../types';

export interface IEventRepository {
  findById(eventId: number): Promise<Event | undefined>;
  findAll(): Promise<Event[]>;
  findByCollege(collegeId: number): Promise<Event[]>;
  findByOrganizer(organizerId: number): Promise<Event[]>;
  create(event: Omit<Event, 'eventId' | 'currentRegistrations' | 'createdAt'>): Promise<Event>;
  update(eventId: number, event: Partial<Event>): Promise<Event | undefined>;
  incrementRegistrations(eventId: number, delta: number): Promise<void>;
  updateStatus(eventId: number, status: EventStatus): Promise<void>;
  delete(eventId: number): Promise<boolean>;
}

export class EventRepository implements IEventRepository {
  private dbManager: DatabaseManager;

  constructor() {
    this.dbManager = DatabaseManager.getInstance();
  }

  public async findById(eventId: number): Promise<Event | undefined> {
    const db = this.dbManager.getDb();
    return db.get<Event>(
      `SELECT e.*, c.name as organizingCollegeName
       FROM EVENT e
       JOIN COLLEGE c ON e.collegeId = c.collegeId
       WHERE e.eventId = ?`,
      [eventId]
    );
  }

  public async findByTitle(title: string): Promise<Event | undefined> {
    const db = this.dbManager.getDb();
    return db.get<Event>(
      `SELECT e.*, c.name as organizingCollegeName
       FROM EVENT e
       JOIN COLLEGE c ON e.collegeId = c.collegeId
       WHERE e.title = ?`,
      [title]
    );
  }

  public async findAll(): Promise<Event[]> {
    const db = this.dbManager.getDb();
    return db.all<Event[]>(
      `SELECT e.*, c.name as organizingCollegeName
       FROM EVENT e
       JOIN COLLEGE c ON e.collegeId = c.collegeId
       ORDER BY e.date ASC, e.startTime ASC`
    );
  }

  public async findByCollege(collegeId: number): Promise<Event[]> {
    const db = this.dbManager.getDb();
    return db.all<Event[]>(
      `SELECT e.*, c.name as organizingCollegeName
       FROM EVENT e
       JOIN COLLEGE c ON e.collegeId = c.collegeId
       WHERE e.collegeId = ?
       ORDER BY e.date ASC, e.startTime ASC`,
      [collegeId]
    );
  }

  public async findByOrganizer(organizerId: number): Promise<Event[]> {
    const db = this.dbManager.getDb();
    return db.all<Event[]>(
      `SELECT e.*, c.name as organizingCollegeName
       FROM EVENT e
       JOIN COLLEGE c ON e.collegeId = c.collegeId
       WHERE e.organizerId = ?
       ORDER BY e.date ASC, e.startTime ASC`,
      [organizerId]
    );
  }

  public async create(event: Omit<Event, 'eventId' | 'currentRegistrations' | 'createdAt'>): Promise<Event> {
    const db = this.dbManager.getDb();
    const result = await db.run(
      `INSERT INTO EVENT (
        title, description, category, eventType, collegeId, organizerId, organizerName,
        date, startTime, endTime, venue, registrationDeadline, capacity, currentRegistrations,
        registrationFee, eligibilityDescription, status, contactInformation
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        event.title,
        event.description,
        event.category,
        event.eventType,
        event.collegeId,
        event.organizerId,
        event.organizerName,
        event.date,
        event.startTime,
        event.endTime,
        event.venue,
        event.registrationDeadline,
        event.capacity,
        0,
        event.registrationFee || 0,
        event.eligibilityDescription,
        event.status || 'PUBLISHED',
        event.contactInformation
      ]
    );

    if (result.lastID) {
      const newEvent = await this.findById(result.lastID);
      if (newEvent) return newEvent;
    }
    const fallback = await this.findByTitle(event.title);
    return fallback!;
  }

  public async update(eventId: number, fields: Partial<Event>): Promise<Event | undefined> {
    const db = this.dbManager.getDb();
    const allowedKeys: (keyof Event)[] = [
      'title', 'description', 'category', 'eventType', 'date', 'startTime', 'endTime',
      'venue', 'registrationDeadline', 'capacity', 'registrationFee',
      'eligibilityDescription', 'status', 'contactInformation'
    ];

    const updates: string[] = [];
    const values: any[] = [];

    for (const key of allowedKeys) {
      if (fields[key] !== undefined) {
        updates.push(`${key} = ?`);
        values.push(fields[key]);
      }
    }

    if (updates.length === 0) {
      return this.findById(eventId);
    }

    values.push(eventId);
    await db.run(`UPDATE EVENT SET ${updates.join(', ')} WHERE eventId = ?`, values);
    return this.findById(eventId);
  }

  public async incrementRegistrations(eventId: number, delta: number): Promise<void> {
    const db = this.dbManager.getDb();
    await db.run(
      `UPDATE EVENT SET currentRegistrations = currentRegistrations + ? WHERE eventId = ?`,
      [delta, eventId]
    );

    const event = await this.findById(eventId);
    if (event && event.currentRegistrations >= event.capacity && event.status === 'REGISTRATION_OPEN') {
      await this.updateStatus(eventId, 'FULL');
    } else if (event && event.currentRegistrations < event.capacity && event.status === 'FULL') {
      await this.updateStatus(eventId, 'REGISTRATION_OPEN');
    }
  }

  public async updateStatus(eventId: number, status: EventStatus): Promise<void> {
    const db = this.dbManager.getDb();
    await db.run('UPDATE EVENT SET status = ? WHERE eventId = ?', [status, eventId]);
  }

  public async delete(eventId: number): Promise<boolean> {
    const db = this.dbManager.getDb();
    const result = await db.run('DELETE FROM EVENT WHERE eventId = ?', [eventId]);
    return (result.changes || 0) > 0;
  }
}

File: backend-java/src/main/java/com/eventbridge/repository/EventRepository.java
package com.eventbridge.repository;

import com.eventbridge.db.DatabaseManager;
import com.eventbridge.model.Event;
import com.eventbridge.patterns.proxy.IEventRepository;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class EventRepository implements IEventRepository {
    private final DatabaseManager dbManager = DatabaseManager.getInstance();

    @Override
    public Event findById(int id) {
        String sql = "SELECT e.*, c.name as organizingCollegeName FROM EVENT e JOIN COLLEGE c ON e.collegeId = c.collegeId WHERE e.eventId = ?";
        try (Connection conn = dbManager.getConnection();
             PreparedStatement pstmt = conn.prepareStatement(sql)) {
            pstmt.setInt(1, id);
            try (ResultSet rs = pstmt.executeQuery()) {
                if (rs.next()) {
                    return mapEvent(rs);
                }
            }
        } catch (SQLException e) {
            System.err.println("Error finding event by ID: " + e.getMessage());
        }
        return null;
    }

    @Override
    public List<Event> findAll() {
        List<Event> list = new ArrayList<>();
        String sql = "SELECT e.*, c.name as organizingCollegeName FROM EVENT e JOIN COLLEGE c ON e.collegeId = c.collegeId ORDER BY e.eventId DESC";
        try (Connection conn = dbManager.getConnection();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(sql)) {
            while (rs.next()) {
                list.add(mapEvent(rs));
            }
        } catch (SQLException e) {
            System.err.println("Error fetching all events: " + e.getMessage());
        }
        return list;
    }

    @Override
    public List<Event> findAllEligibleForUser(com.eventbridge.model.User user) {
        return findAll();
    }

    public void incrementRegistrationCount(int eventId) {
        String sql = "UPDATE EVENT SET currentRegistrations = currentRegistrations + 1 WHERE eventId = ?";
        try (Connection conn = dbManager.getConnection();
             PreparedStatement pstmt = conn.prepareStatement(sql)) {
            pstmt.setInt(1, eventId);
            pstmt.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error updating registration count: " + e.getMessage());
        }
    }

    private Event mapEvent(ResultSet rs) throws SQLException {
        Event e = new Event();
        e.setEventId(rs.getInt("eventId"));
        e.setTitle(rs.getString("title"));
        e.setDescription(rs.getString("description"));
        e.setCategory(rs.getString("category"));
        e.setEventType(rs.getString("eventType"));
        e.setCollegeId(rs.getInt("collegeId"));
        e.setOrganizingCollegeName(rs.getString("organizingCollegeName"));
        e.setOrganizerId(rs.getInt("organizerId"));
        e.setOrganizerName(rs.getString("organizerName"));
        e.setDate(rs.getString("date"));
        e.setStartTime(rs.getString("startTime"));
        e.setEndTime(rs.getString("endTime"));
        e.setVenue(rs.getString("venue"));
        e.setRegistrationDeadline(rs.getString("registrationDeadline"));
        e.setCapacity(rs.getInt("capacity"));
        e.setCurrentRegistrations(rs.getInt("currentRegistrations"));
        e.setRegistrationFee(rs.getDouble("registrationFee"));
        e.setEligibilityDescription(rs.getString("eligibilityDescription"));
        e.setStatus(rs.getString("status"));
        e.setContactInformation(rs.getString("contactInformation"));
        return e;
    }
}
